### PR TITLE
Treatment SQL performs well on multi-study queries

### LINKED ISSUE #8063: Several count API endpoints very slow on v3.4.21 and v3.4.22
Try e.g. https://master.cbioportal.org/study/summary?id=5dd408f0e4b0f7d2de7862a8. A lot of endpoints take over 40 seconds:

Screen Shot 2020-11-16 at 10 29 19 AM

In production, which is running v3.4.20, the slowest endpoint is gene panel data:

Screen Shot 2020-11-16 at 10 30 36 AM

### PR DESCRIPTION
**Pull request by Luke-Sikina**
Fix #8063

Describe changes proposed in this pull request:
- Filter cancer study before filtering by study, sample id pairs

### CODE REVIEW

File: persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/TreatmentMapper.java [hunk 1]
```diff
@@ -7,13 +7,13 @@
...
 
 public interface TreatmentMapper {
-    List<Treatment> getAllTreatments(List<String> sampleIds, List<String> studyIds);
+    List<Treatment> getAllTreatments(List<String> sampleIds, List<String> studyIds, List<String> uniqueStudyIds);
 
-    List<ClinicalEventSample> getAllSamples(List<String> sampleIds, List<String> studyIds);
+    List<ClinicalEventSample> getAllSamples(List<String> sampleIds, List<String> studyIds, List<String> uniqueStudyIds);
 
-    Set<String> getAllUniqueTreatments(List<String> sampleIds, List<String> studyIds);
+    Set<String> getAllUniqueTreatments(List<String> sampleIds, List<String> studyIds, List<String> uniqueStudyIds);
```
Comment (inodb) [inline persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/TreatmentMapper.java:14]: might be cleaner if the unique/distinct step can be done in mysql in some way. Looks slightly odd when studyIds are given as argument twice imo

**Review by inodb — APPROVED** [review]
Thanks! This seems like it could work (didn't fully test, but easier to try in master once merged)

I wonder if we should also change in the frontend how we load these counts given that all rendering/display currently waits for all data to be loaded. Maybe with these fixes the latency is minor, but just curious if u had any thoughts on that @Luke-Sikina

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/TreatmentMapper.xml [hunk 2]
```diff
@@ -99,7 +99,11 @@
                     </foreach>
                 </if>
                 <if test="@java.util.Arrays@stream(studyIds.toArray()).distinct().count() > 1">
-                    (cancer_study.CANCER_STUDY_IDENTIFIER, sample.STABLE_ID) IN
+                    cancer_study.CANCER_STUDY_IDENTIFIER IN
+                    <foreach item="item" collection="uniqueStudyIds" open="(" separator="," close=")">
```
Comment (kalletlak) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/TreatmentMapper.xml:103]: Could you try this and see if it works
```
 <foreach item="item" collection="@java.util.Arrays@stream(studyIds.toArray()).distinct().collect(@java.util.stream.Collectors.toList())" open="(" separator="," close=")">
    #{item}
 </foreach>
```

Comment (Luke-Sikina) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/TreatmentMapper.xml]: This also works, and this fixes Ino's concerns. Updated to match.

**Review by kalletlak — APPROVED** [review]

**Pull request merged**